syntax = "proto3";

enum CalculationMode {
    CPU = 0;
    GPU = 1;
}

message Position {
    float x = 1;
    float y = 2;
    float z = 3;
}

message Velocity {
    float x = 1;
    float y = 2;
    float z = 3;
}

message Body {
    string name = 1;
    Position position = 2;
    Velocity velocity = 3;
    float weight = 4;
}

message NBodyProblemData {
    float dt = 1;
    int32 num_iterations = 2;
    float G = 3;
    repeated Body bodies = 4;
    CalculationMode calculation_mode = 5;
}

message OneIterationSolutionNBodyProblem {
    int32 num_iteration = 1;
    repeated Body bodies = 2;
}

message NBodyProblemSolution {
    repeated OneIterationSolutionNBodyProblem iterations = 1;
}

service NBodyProblemSolver {
    rpc solve(NBodyProblemData) returns (NBodyProblemSolution) { }
}

